Sessions sync

bdragan opened the pull request:
### Session Synchronization with Remote Storage

This pull request introduces a comprehensive session synchronization feature, enabling users to back up, share, and synchronize their work sessions across different machines via a remote storage service.

#### Summary of Changes

The core of this feature is the ability to treat a remote server as a source of truth for sessions associated with a project. Local changes are uploaded, remote changes are downloaded, and the local session list is kept in sync with the remote. This provides a seamless experience for users working on the same project from multiple locations.

#### Key Functionalities

*   **Automatic Upload of Local Changes:**
    *   Any new session, modification (e.g., a new conversation step), or rename is automatically marked as "dirty" and queued for upload.
    *   On application startup and periodically in the background (every 30 seconds), the client uploads these pending changes to the remote server.
    *   When the application is closed, a final attempt is made to upload any remaining unsynced changes.

*   **Synchronization from Remote:**
    *   On startup, after uploading local changes, the application fetches the list of all sessions for the project from the remote server.
    *   It downloads any sessions that exist remotely but not locally.
    *   It updates local sessions if a newer version is available on the server (based on modification timestamps).
    *   It removes local sessions that have been deleted from the remote, ensuring the local view doesn't retain stale data.

*   **Conflict-Free Deletion:**
    *   When a session is deleted locally, it is not immediately removed. Instead, it is moved to a temporary "deleted" location and marked as `PENDING_DELETION`.
    *   The background sync process informs the server of the deletion. Once the server confirms, the local copy is marked as fully `DELETED`. This prevents a session deleted locally from being re-downloaded from the remote before the deletion is synced.

*   **On-Demand Updates:**
    *   Before a session is loaded (e.g., when a user switches to it in the UI), the application performs a quick check to see if a more recent version exists on the server based on the cached remote sessions metadata. If so, it downloads the latest version before loading it, ensuring the user works with the most up-to-date session content before background sync process is complete.
    * This is useful during initial loading of active sessions and when users switch to other sessions before background sync is complete to prevent session content divergence (either automatic with `Loaded external changes` and similar, or by users modifying sessions before sync is complete). 
    * This does not lead to degraded performance every time we load session history because we check the session info against cached remote metadata.

#### Workflow Overview

1.  **Startup:**
    *   The application first uploads any local sessions that have been created or modified since the last sync.
    *   It then synchronizes with the remote, downloading new/updated sessions and processing deletions.
    *   A background task is scheduled to periodically upload local changes.

2.  **During Use:**
    *   Creating, renaming, or interacting with a session marks it for the next upload cycle.
    *   Deleting a session marks it for remote deletion.
    *   Switching to a session may trigger a download if a newer version is found remotely.

3.  **Shutdown:**
    *   A final upload of all pending local changes is attempted to ensure no work is lost.

#### Technical Implementation Notes

*   A `SyncStatus` (`PENDING_SYNC`, `SYNCED`, `PENDING_DELETION`, `DELETED`) has been added to the session metadata (`SessionInfo`) to track the state of each session relative to the remote.
*   The `Service` class now includes methods for interacting with the new `/api/sessions` backend endpoints (list, get, write, delete).
*   The `SessionManager` contains the primary logic for handling the synchronization state machine, managing uploads, downloads, and deletions.
*   Projects are identified on the remote server primarily by their Git remote URL, falling back to the project's root directory name. This is exposed via a new `getRemoteProjectName()` method on the `IProject` interface.

jbellis commented:
Sorry for the slow review. I kept getting lost in the details.  Today I figured out the big picture of what bothers me.

My prompt:
> ISTM that the big problem is that there's two paths for a lot of work: the "sync from cold" and "my local queue of stuff"

> simplicity is reliability. i'd prefer to unify these to avoid corner case bugs. the new design should use cached information from a prior sync to save work and time, while keeping it a single unified path for sync activity.

> therefore, syncstate must die (the one piece we need to keep -- in a different form -- is a tombstone special value)

> what else can be simplified away?

GPT5: https://gist.github.com/jbellis/4afcde6e31ad2e831e24c4fb1a20c3dc

jbellis commented on app/src/main/java/io/github/jbellis/brokk/MainProject.java:
```diff
@@ -51,6 +50,7 @@ public final class MainProject extends AbstractProject {
     private final Path styleGuidePath;
     private final Path reviewGuidePath;
     private final SessionManager sessionManager;
+    private final SessionRegistry sessionRegistry = new SessionRegistry();
```
> the same session can be active in multiple main projects in the same JVM 

this doesn't make sense to me at all, any given Session should only be active in at most a single Project at a time

bdragan replied:
I’m re-reading now how I explained this, and I think I understand where the confusion comes from. I don’t mean that the same ‘physical’ session can be active in multiple main projects in the same JVM, but two copies of the same original session (with the same session id) can:
- A session can be copied manually into another project that gets opened in the same JVM afterwards.
- Or it can now be automatically synced into another open main project if the other project has the same remote project name (the same remote url or the same root folder name if not under git).

If we use global `SessionRegistry` with static state, it would e.g. prevent the session from being deleted in one project because we think it is open in another worktree (we just check the static cache), while it is open in completely different project, and changing/deleting it in one project does not affect the copy in the other one in any way. 

Basically, everything else is separate between the projects (including separate session dirs and files, chromes, main project instances, etc), so I also separated session registries in order not to deal with cases mentioned above (plus I would have to deal with it during session sync too because we’d ask user whether to overwrite a session that might not be open at all in the current project, hence I did it here).

bdragan commented:
> The big rock remaining is improving encapsulation so we can test it and make it easier to reason about. Here's one idea (#collaborating)
> 
> 1. instead of a single "synchronizeRemoteSessions" call that prompts user inline, a "getSyncPlan()" call that has placeholders for user decisions, CM gets the decisions, then getSyncPlan(decisions) again (repeat until all required decisions are included, no placeholders), then CM calls "executeSync(plan)".
> 2. Probably getSyncPlan/executeSync should live in a new SessionSync class instead of SessionManager to keep api boundaries clean
> 3. (1) + (2) this would also allow testing getSyncPlan in isolation which is something we should try to do. Need to allow specifying getSyncPlan(localData, remoteData) ofc instead of hardcoding to disk and network.

Indeed, extracted now synchronization logic into a separate class and added tests. As for the approach, probably your suggestion is more flexible, but a sync plan could change from the moment we generate it to the moment we execute it on a session, because sync might take a longer time for multiple sessions (especially if they are large) and the user might have already modified or opened the session in the meantime. So we’d need to abort and regenerate the plan, etc.

I think this way it is easier – once we synced a session (upload/download/delete), we’re done with it until the next sync cycle, and we move on to the next session. Wdyt?

jbellis commented:
Tactical details:

1. It looks like if deleteRemoteSession throws we still delete the local tombstone which is bad
2. clean up all the `catch (Exception)` noise (I can't tell at a glance which should be a more specific exception type and which should just be deleted)
3. removing the session release from GitRepoWorktrees makes me nervous because now all callers need to understand session lifetime; e.g. it looks like the release call is missing from performMergeOperation. Not sure there is a good fix for this other than documenting it, but maybe you can see a better solution.
